server/src/main/java/io/crate/planner/optimizer/joinorder/JoinGraph.java

```diff
@@ -0,0 +1,271 @@
+            }
+            return new JoinGraph(logicalPlan, List.of(logicalPlan), Map.of(), List.of(), false);
+        }
+
+        @Override
+        public JoinGraph visitRename(Rename rename, Context context) {
+            context.resolver.put(rename.name(), rename);
+            return rename.source().accept(this, context);
+        }
+
+        @Override
+        public JoinGraph visitGroupReference(GroupReference groupReference, Context context) {
+            return resolvePlan.apply(groupReference).accept(this, context);
+        }
+
+        @Override
+        public JoinGraph visitFilter(Filter filter, Context context) {
+            var source = filter.source().accept(this, context);
+            return source.withFilters(List.of(filter.query()));
+        }
+
+        @Override
+        public JoinGraph visitJoinPlan(JoinPlan joinPlan, Context context) {
+
+            var left = joinPlan.lhs().accept(this, context);
+            var right = joinPlan.rhs().accept(this, context);
+
+            if (joinPlan.joinType() == JoinType.CROSS) {
+                return left.joinWith(joinPlan, right).withCrossJoin(true);
+            }
+
+            if (joinPlan.joinType() != JoinType.INNER) {
+                return left.joinWith(joinPlan, right);
+            }
+
+            var joinCondition = joinPlan.joinCondition();
+            var edgeCollector = new EdgeCollector();
+            if (joinCondition != null) {
+                var split = QuerySplitter.split(joinCondition);
+                for (var entry : split.entrySet()) {
+                    if (entry.getKey().size() == 2) {
+                        entry.getValue().accept(edgeCollector, context);
+                    }
+                }
+            }
+            return left.joinWith(joinPlan, right).withEdges(edgeCollector.edges);
+        }
+
+        private static class EdgeCollector extends SymbolVisitor<Context, Void> {
+
+            private final Map<LogicalPlan, Set<Edge>> edges = new HashMap<>();
+            private final List<LogicalPlan> sources = new ArrayList<>();
+
+            @Override
+            public Void visitField(ScopedSymbol s, Context context) {
+                var fieldResolver = context.resolver.get(s.relation());
+                var resolved = fieldResolver.resolveField(s);
+                sources.add(context.referencesToRelations.get(resolved));
+                return null;
+            }
+
+            @Override
+            public Void visitReference(Reference ref, Context context) {
+                sources.add(context.referencesToRelations.get(ref));
+                return null;
+            }
+
+            @Override
+            public Void visitFunction(io.crate.expression.symbol.Function f, Context context) {
+                f.arguments().forEach(x -> x.accept(this, context));
+                if (f.name().equals(EqOperator.NAME)) {
+                    var fromSymbol = f.arguments().get(0);
```

Comment:
This would break with literals/non join conditions - I guess that's fine given how it's used but it would help to state the contract explicitly. E.g. Maybe add an assertion before that the size of sources increased by 2 after `f.arguments().forEach(x -> x.accept(this, context));` with a clear message